src/Watch.cpp

```diff
@@ -308,8 +308,158 @@ void Watch::parseWO(WatchInfo& wi, int first, int count) {
+    time_t wo_start = mktime(&wo.start_time.time);
+
+    for (unsigned int lap = 0; lap < wo.lapcount; lap++) {
+        // Initialise storage
+        wo.lapinfo[lap].split = tm();
+        wo.lapinfo[lap].split.tm_year = 70;
+        wo.lapinfo[lap].split.tm_mday = 1;
+        wo.lapinfo[lap].split_milli = 0;
+
+        // 1-3 time
+        wo.lapinfo[lap].split.tm_hour = *it++;
+        wo.lapinfo[lap].split.tm_min = *it++;
+        wo.lapinfo[lap].split.tm_sec = *it++;
+
+        // 4 (milli) is expected to be printed as hex. eg data 83 = 0x53 displayed
+        std::ostringstream hexstream;
+        hexstream << std::hex << (int)*it++;
+        wo.lapinfo[lap].split_milli = atoi(hexstream.str().c_str());
+
+        // Calculate lap time
+        time_t this_split;
+        time_t prev_split;
+        time_t this_lap;
+        time_t abs_split;
+        time_t start_t;
+
+        int prev_milli;
+        int this_milli;
+
+        wo.lapinfo[lap].lap = tm();
+        wo.lapinfo[lap].lap.tm_year = 70;
+        wo.lapinfo[lap].lap.tm_mday = 1;
+        wo.lapinfo[lap].lap_milli = 0;
+
+        if (lap == 0) {
+            prev_split = 0;
+            prev_milli = 0;
+        }
+        else {
+            prev_split = my_timegm(&wo.lapinfo[lap - 1].split);
+            prev_milli = wo.lapinfo[lap - 1].split_milli;
+        }
+
+        this_split = my_timegm(&wo.lapinfo[lap].split);
+        this_lap = this_split - prev_split;
+        this_milli = wo.lapinfo[lap].split_milli - prev_milli;
+
+        while (this_milli < 0) {
+            this_milli += 100;
+            this_lap--;
+        }
+
+        wo.lapinfo[lap].lap = *gmtime(&this_lap);
+        wo.lapinfo[lap].lap_milli = this_milli;
+        wo.lapinfo[lap].lap_seconds = (int) this_lap;
+
+        start_t = wo_start + prev_split;
+        abs_split = wo_start + this_split + 1;
+
+        wo.lapinfo[lap].abs_split.time = *localtime(&abs_split);
+        wo.lapinfo[lap].start_time.time = *localtime(&start_t);
+
+        // 5 HR
+        wo.lapinfo[lap].avg_hr = *it++;
+
+        // 6-8 all zero?
+        it += 3;
+
+        // 9 lap dist m
+        // 10 lap dist * 256m
+        // 11-12 etc?
+        wo.lapinfo[lap].distance = *it++;
+        wo.lapinfo[lap].distance += *it++ << 8;
+        wo.lapinfo[lap].distance += *it++ << 16;
+        wo.lapinfo[lap].distance += *it++ << 24;
+
+        // 13 lap speed, in hundreds of m/hr - divide by 10 for km/h
+        wo.lapinfo[lap].speed = *it++;
+        wo.lapinfo[lap].speed += *it++ << 8;
+        wo.lapinfo[lap].speed += *it++ << 16;
+        wo.lapinfo[lap].speed += *it++ << 24;
+
+        double speed = wo.lapinfo[lap].speed / 10.0;
+        int pacemin;
+        int pacesec;
+
+        pacemin = 60 / speed;
+        pacesec = (60.0 / speed - pacemin) * 60;
+        wo.lapinfo[lap].pace = tm();
+        wo.lapinfo[lap].pace.tm_min = pacemin;
+        wo.lapinfo[lap].pace.tm_sec = pacesec;
+
+        if(pacemin > 39) {
```

Comment:
I was fiddling the value so that it matched the output of the watch itself
but wasn't sure if it was actually a good idea to force the loss of
resolution into the data structure itself.

The rounding to 40:00/km is actually pretty slow so perhaps it is worth
capping at that?

On Tue Feb 17 2015 at 01:27:48 mru <notifications@github.com> wrote:

> In src/Watch.cpp
> <https://github.com/mru00/crane_gps_watch/pull/10#discussion_r24753254>:
>
> > +        wo.lapinfo[lap].speed = *it++;
> > +        wo.lapinfo[lap].speed += *it++ << 8;
> > +        wo.lapinfo[lap].speed += *it++ << 16;
> > +        wo.lapinfo[lap].speed += *it++ << 24;
> > +
> > +        double speed = wo.lapinfo[lap].speed / 10.0;
> > +        int pacemin;
> > +        int pacesec;
> > +
> > +        pacemin = 60 / speed;
> > +        pacesec = (60.0 / speed - pacemin) * 60;
> > +        wo.lapinfo[lap].pace = tm();
> > +        wo.lapinfo[lap].pace.tm_min = pacemin;
> > +        wo.lapinfo[lap].pace.tm_sec = pacesec;
> > +
> > +        if(pacemin > 39) {
>
> should this saturation also apply for lapinfo.pace?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mru00/crane_gps_watch/pull/10/files#r24753254>.
>